transport/TransportImpl.cpp

```diff
@@ -1969,17 +1989,28 @@ void TransportImpl::onIceStateChanged(ice::IceSession* session, const ice::IceSe
             }
         }
 
-        while (!_rtpEndpoints.empty() && _rtpEndpoints.back().get() != _selectedRtp &&
-            _rtpEndpoints.back()->getTransportType() == ice::TransportType::TCP)
-        {
-            _rtpEndpoints.back()->unregisterListener(this);
-            _rtpEndpoints.pop_back();
-        }
-        while (!_rtpEndpoints.empty() && _rtpEndpoints.front().get() != _selectedRtp &&
-            _rtpEndpoints.front()->getTransportType() == ice::TransportType::TCP)
+        if (_rtpIceSession->getRole() == ice::IceRole::CONTROLLING)
         {
-            _rtpEndpoints.front()->unregisterListener(this);
-            _rtpEndpoints.erase(_rtpEndpoints.begin());
+            while (!_rtpEndpoints.empty() && _rtpEndpoints.back().get() != _selectedRtp &&
+                _rtpEndpoints.back()->getTransportType() == ice::TransportType::TCP)
+            {
+                logger::info("discarding %s, ref %ld",
+                    _loggableId.c_str(),
+                    _rtpEndpoints.back()->getName(),
+                    _rtpEndpoints.back().use_count());
+                _rtpIceSession->onTcpRemoved(_rtpEndpoints.back().get());
```

Comment:
This is the crash fix. To remove the endpoint from ice session before it is actually removed by transport factory.